pil-analyzer/src/statement_processor.rs

```diff
@@ -577,20 +613,48 @@ where
     }
 
     fn process_enum_declaration(
-        &self,
+        &mut self,
+        source: SourceRef,
+        name: String,
+        symbol: Symbol,
         enum_decl: EnumDeclaration<parsed::Expression>,
-    ) -> EnumDeclaration {
+    ) -> Vec<PILItem> {
         let type_vars = enum_decl.type_vars.vars().collect();
         let variants = enum_decl
             .variants
             .into_iter()
             .map(|v| self.process_enum_variant(v, &type_vars))
             .collect();
-        EnumDeclaration {
+        let enum_decl = EnumDeclaration {
             name: self.driver.resolve_decl(&enum_decl.name),
             type_vars: enum_decl.type_vars,
             variants,
-        }
+        };
+
+        let shared_enum_decl = Arc::new(enum_decl.clone());
+
+        let inner_items = enum_decl
+            .variants
+            .iter()
+            .map(|variant| {
+                (
+                    vec![name.clone(), variant.name.clone()],
+                    FunctionValueDefinition::TypeConstructor(
+                        shared_enum_decl.clone(),
+                        variant.clone(),
+                    ),
+                )
+            })
+            .collect();
```

Comment:
Does this need to be a vector or can it be an iterator?

Reply:
I don't think so, but at least I simplify everything around to try to avoid some clones and make it cleaner